### PR TITLE
Refactored Wishart and MatrixNormal distribution

### PR DESCRIPTION
**Pull request by kc611**
This PR refactors the `Wishart` and `MatrixNormal` multivariate distributions.

### CODE REVIEW

File: pymc3/distributions/multivariate.py [hunk 1]
```diff
@@ -1531,175 +1560,96 @@ class MatrixNormal(Continuous):
...
+    def dist(
+        cls,
+        mu,
         rowcov=None,
         rowchol=None,
-        rowtau=None,
         colcov=None,
         colchol=None,
-        coltau=None,
         shape=None,
```
Comment (kc611) [inline pymc3/distributions/multivariate.py:1579]: The `rowtau` and `choltau` arguments are not supported right now. Maybe there was a plan for supporting those (but was maybe abandoned midway?).

File: pymc3/tests/test_distributions.py [hunk 2]
```diff
@@ -1893,8 +1893,7 @@ def test_mvnormal_init_fail(self):
             with pytest.raises(ValueError):
                 x = MvNormal("x", mu=np.zeros(3), cov=np.eye(3), tau=np.eye(3), size=3)
 
-    @pytest.mark.parametrize("n", [1, 2, 3])
-    @pytest.mark.xfail(reason="Distribution not refactored yet")
+    @pytest.mark.parametrize("n", [2, 3])
```
Comment (kc611) [inline pymc3/tests/test_distributions.py:1896]: This fails for n=1 , maybe the dimensions are being downcasted to a scalar when matrix is shaped `1x1`.

File: pymc3/tests/test_distributions_random.py [hunk 3]
```diff
@@ -1306,6 +1303,56 @@ class TestOrderedProbit(BaseTestDistribution):
...
+    reference_dist = lambda self: functools.partial(
+        self.matrixnormal_rng_fn, rng=self.get_random_state()
+    )
+    sizes_to_check = [None, (1), (2, 3)]
+    sizes_expected = [(3,), (1, 3), (2, 3, 3)]
+
+    tests_to_run = [
+        "check_rv_size",
+        "check_pymc_params_match_rv_op",
+        "check_pymc_draws_match_reference",
```
Comment (kc611) [inline pymc3/tests/test_distributions_random.py:1352]: This test fails, strangely the first element matches everytime. @brandonwillard is this related to https://github.com/aesara-devs/aesara/issues/209 ? There seems to a similar operation involved here.

File: pymc3/tests/test_distributions_random.py [hunk 4]
```diff
@@ -1306,6 +1304,71 @@ class TestOrderedProbit(BaseTestDistribution):
...
+        (mu,) = broadcast_dist_samples_to(to_shape=output_shape, samples=[mu], size=size)
+        rowchol = np.broadcast_to(rowchol, shape=size + rowchol.shape[-2:])
+
+        colchol = np.broadcast_to(colchol, shape=size + colchol.shape[-2:])
+        colchol = np.swapaxes(colchol, -1, -2)  # Take transpose
+
+        standard_normal = rng.standard_normal(output_shape)
+        samples = mu + np.matmul(rowchol, np.matmul(standard_normal, colchol))
+
+        return samples
```
Comment (kc611) [inline pymc3/tests/test_distributions_random.py]: Kind-of counter-intuitive to use this function here as reference, since it's basically same as the internal one, but it looks like `st.matrix_normal.rvs(mean=mu, rowcov=rowcov, colcov=colcov, size=size)` used previously resulted in draws that are differently valued than our implementation and our previous testing just checked if the values came from similar distributions. (See previous commit https://github.com/pymc-devs/pymc3/pull/4777/commits/4ee1360158885556c7801b97bc166ae2a40c37bd).

File: pymc3/distributions/multivariate.py (hunk 1 above)
Comment (brandonwillard) [reply pymc3/distributions/multivariate.py:1579]: They weren't supported in `v3`?

Comment (kc611) [reply pymc3/distributions/multivariate.py:1579]: No, passing either of `rowtau` or `choltau` raised an error. (It's in the code right below this)

File: pymc3/tests/test_distributions_random.py (hunk 4 above)
Comment (ricardoV94) [reply pymc3/tests/test_distributions_random.py]: Does this help? https://github.com/pymc-devs/pymc3/blob/fe115b857ae42e84d1f359153fd1f7efb2c327ae/docs/source/distribution_implementation.md#notes-on-check_pymcs_draws_match_reference-test

Comment (ricardoV94) [reply pymc3/tests/test_distributions_random.py]: In this case, it may be better to use something along the lines of the `pymc3_random` test and compare with the Scipy version. Maybe compare dimension by dimension since it is a multivariate distribution.

Comment (kc611) [reply pymc3/tests/test_distributions_random.py]: In that case we can add a seperate class which does this. (Similar to how we did with `Interpolated` distribution )

File: pymc3/tests/test_distributions.py (hunk 2 above)
Comment (brandonwillard) [reply pymc3/tests/test_distributions.py:1896]: If that `n` is being used to determine the number of output dimensions, then `n=1` might not make sense for a matrix-valued distribution like this.

File: pymc3/tests/test_distributions_random.py (hunk 4 above)
Comment (ricardoV94) [reply pymc3/tests/test_distributions_random.py]: You can also add it in that class to keep things organized, such as this: https://github.com/pymc-devs/pymc3/blob/f4343ca7ba112f843cf256378e88ff80a25342a6/pymc3/tests/test_distributions_random.py#L385-L390

File: pymc3/distributions/multivariate.py [hunk 5]
```diff
@@ -737,6 +738,21 @@ def __str__(self):
 matrix_pos_def = PosDefMatrix()
 
 
+class WishartRV(RandomVariable):
+    name = "wishart"
+    ndim_supp = 0
```
Comment (brandonwillard) [inline pymc3/distributions/multivariate.py]: Shouldn't this be `ndim_supp = 2` (i.e. the Wishart distribution is over matrices)?

File: pymc3/distributions/multivariate.py [hunk 6]
```diff
@@ -785,34 +802,15 @@ def __init__(self, nu, V, *args, **kwargs):
...
-        Parameters
-        ----------
-        point: dict, optional
-            Dict of variable values on which random values are to be
-            conditioned (uses default point if not specified).
-        size: int, optional
-            Desired size of random sample (returns one sample if not
-            specified).
+        nu = at.as_tensor_variable(nu)
+        V = at.as_tensor_variable(V)
```
Comment (ricardoV94) [inline pymc3/distributions/multivariate.py]:
```suggestion
        nu = at.as_tensor_variable(intX(nu))
        V = at.as_tensor_variable(floatX(V))
```

File: pymc3/tests/test_distributions.py (hunk 2 above)
Comment (kc611) [reply pymc3/tests/test_distributions.py:1896]: I raised an error if such a matrix is passed.

File: pymc3/distributions/multivariate.py [hunk 7]
```diff
@@ -1531,175 +1558,101 @@ class MatrixNormal(Continuous):
...
         else:
-            self.n = colchol.shape[0]
-            self._colcov_type = "chol"
             if colchol.ndim != 2:
                 raise ValueError("colchol must be two dimensional.")
-            self.colchol_cov = at.as_tensor_variable(colchol)
+            colchol_cov = at.as_tensor_variable(colchol)
 
-    def random(self, point=None, size=None):
+        mu = at.as_tensor_variable(mu)
```
Comment (ricardoV94) [inline pymc3/distributions/multivariate.py]:
```suggestion
        mu = at.as_tensor_variable(floatX(mu))
```

File: pymc3/tests/test_distributions.py [hunk 8]
```diff
@@ -1907,6 +1906,8 @@ def test_matrixnormal(self, n):
                 "colcov": PdMatrix(n) * mat_scale,
             },
             matrix_normal_logpdf_cov,
+            extra_args={"size": n},
+            decimal=select_by_precision(float64=5, float32=5),
```
Comment (ricardoV94) [inline pymc3/tests/test_distributions.py]: Do the tests pass if aesara `floatX` is set to "float32"? Usually the default precision we use for float32 is `3`. Your change is increasing the precision compared to what we were testing before.

Comment (ricardoV94) [reply pymc3/tests/test_distributions.py]: Also make sure you test all combinations locally by setting `n_samples=-1` (and test both in float32 and float64)

File: pymc3/distributions/multivariate.py (hunk 5 above)
Comment (kc611) [reply pymc3/distributions/multivariate.py]: It somehow fails if we set `ndim_supp = 2`. Strangely, the samples returned by `rng_fn` are actually 2-dimensional. Any ideas why this might be happening.

File: pymc3/tests/test_distributions.py (hunk 8 above)
Comment (kc611) [reply pymc3/tests/test_distributions.py]: I set the precision to 2 decimals for `float32` just to be sure.

Comment (ricardoV94) [reply pymc3/tests/test_distributions.py]: You should actually test it locally. 2 might not be necessary or it might not be enough.

Comment (ricardoV94) [reply pymc3/tests/test_distributions.py]: You can check the notes 3 and 4 in here: https://github.com/pymc-devs/pymc3/blob/fe115b857ae42e84d1f359153fd1f7efb2c327ae/docs/source/distribution_implementation.md#4-adding-tests-for-the-logp--logcdf-methods

File: pymc3/distributions/multivariate.py (hunk 5 above)
Comment (ricardoV94) [reply pymc3/distributions/multivariate.py]: You might need to override `_shape_from_params` if the RV is failing to guess the shape with the default method. This is needed here for example: https://github.com/aesara-devs/aesara/blob/3a3adaee3b8ceccff6d0208996f117fcd5e5eb6f/aesara/tensor/random/basic.py#L518-L521

What error are you seeing?

Comment (kc611) [reply pymc3/distributions/multivariate.py]: > What error are you seeing?

<details>
<summary>Here's the full traceback</summary>
<br>





```
________________________________________________________ TestWishart.test_distribution ________________________________________________________

self = <pymc3.tests.test_distributions_random.TestWishart object at 0x7f53e0b50810>

    def test_distribution(self):
        self.validate_tests_list()
>       self._instantiate_pymc_rv()

pymc3/tests/test_distributions_random.py:300: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pymc3/tests/test_distributions_random.py:311: in _instantiate_pymc_rv
    **params, size=self.size, rng=aesara.shared(self.get_random_state(reset=True))
pymc3/distributions/multivariate.py:816: in dist
    return super().dist([nu, V], *args, **kwargs)
pymc3/distributions/distribution.py:285: in dist
    rv_out = cls.rv_op(*dist_params, size=create_size, **kwargs)
../../../miniconda3/envs/pymc3-dev-py37/lib/python3.7/site-packages/aesara/tensor/random/op.py:323: in __call__
    res = super().__call__(rng, size, dtype, *args, **kwargs)
../../../miniconda3/envs/pymc3-dev-py37/lib/python3.7/site-packages/aesara/graph/op.py:269: in __call__
    node = self.make_node(*inputs, **kwargs)
../../../miniconda3/envs/pymc3-dev-py37/lib/python3.7/site-packages/aesara/tensor/random/op.py:370: in make_node
    bcast = self.compute_bcast(dist_params, size)
../../../miniconda3/envs/pymc3-dev-py37/lib/python3.7/site-packages/aesara/configparser.py:49: in res
    return f(*args, **kwargs)
../../../miniconda3/envs/pymc3-dev-py37/lib/python3.7/site-packages/aesara/tensor/random/op.py:287: in compute_bcast
    shape = self._infer_shape(size, dist_params)
../../../miniconda3/envs/pymc3-dev-py37/lib/python3.7/site-packages/aesara/tensor/random/op.py:257: in _infer_shape
    param_shapes=param_shapes,
pymc3/distributions/multivariate.py:750: in _shape_from_params
    self.ndim_supp, dist_params, rep_param_idx, param_shapes
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

ndim_supp = 2, dist_params = (TensorConstant{4}, TensorConstant{[[1. 0. 0...0. 0. 1.]]}), rep_param_idx = 2
param_shapes = [(), (Subtensor{int64}.0, Subtensor{int64}.0)]

    def default_shape_from_params(
        ndim_supp, dist_params, rep_param_idx=0, param_shapes=None
    ):
        """Infer the dimensions for the output of a `RandomVariable`.
    
        This is a function that derives a random variable's support
        shape/dimensions from one of its parameters.
    
        XXX: It's not always possible to determine a random variable's support
        shape from its parameters, so this function has fundamentally limited
        applicability and must be replaced by custom logic in such cases.
    
        XXX: This function is not expected to handle `ndim_supp = 0` (i.e.
        scalars), since that is already definitively handled in the `Op` that
        calls this.
    
        TODO: Consider using `aesara.compile.ops.shape_i` alongside `ShapeFeature`.
    
        Parameters
        ----------
        ndim_supp: int
            Total number of dimensions for a single draw of the random variable
            (e.g. a multivariate normal draw is 1D, so `ndim_supp = 1`).
        dist_params: list of `aesara.graph.basic.Variable`
            The distribution parameters.
        param_shapes: list of tuple of `ScalarVariable` (optional)
            Symbolic shapes for each distribution parameter.  These will
            be used in place of distribution parameter-generated shapes.
        rep_param_idx: int (optional)
            The index of the distribution parameter to use as a reference
            In other words, a parameter in `dist_param` with a shape corresponding
            to the support's shape.
            The default is the first parameter (i.e. the value 0).
    
        Results
        -------
        out: a tuple representing the support shape for a distribution with the
        given `dist_params`.
    
        """
        if ndim_supp <= 0:
            raise ValueError("ndim_supp must be greater than 0")
        if param_shapes is not None:
>           ref_param = param_shapes[rep_param_idx]
E           IndexError: list index out of range

../../../miniconda3/envs/pymc3-dev-py37/lib/python3.7/site-packages/aesara/tensor/random/op.py:69: IndexError
```
</details>

And overriding `_shape_from_params` didn't work. You might also wanna take a look at the test first, just to see if the parameters are valid.

Comment (ricardoV94) [reply pymc3/distributions/multivariate.py]: It's definitely looks like you need to override that method. I didn't mean to imply that it should be the same overwrite, just that it needs one.

Comment (brandonwillard) [reply pymc3/distributions/multivariate.py]: > It somehow fails if we set `ndim_supp = 2`. Strangely, the samples returned by `rng_fn` are actually 2-dimensional. Any ideas why this might be happening?

`ndim_supp` is used for shape inference and a few other related things, but not necessarily for critical `RandomVariable.perform` operations, so it's not surprising that this value would be incorrect and still produce valid samples.

Regardless, we need the `ndim_supp` values to be accurate.  We might also need to add some more sanity checks at the symbolic level using that property (i.e. when/if we can verify that the output dimensions match `ndim_supp`, we probably should).

For example, we might be missing a sanity check like `outputs[1].ndim >= self.ndim_supp` in `RandomVariable.make_node`.  It would be better if we could somehow guarantee that condition from the construction of `outputs`, but given how subclasses can override some of the steps involved, such a simple check might be worthwhile.

Comment (brandonwillard) [reply pymc3/distributions/multivariate.py]: Otherwise, @ricardoV94 is correct; it looks like this needs a custom `_shape_from_params` that says the second parameter, `V`, determines the output shape.  I think the default assumes that the first parameter does, which is clearly wrong in this case.

**Review by brandonwillard — CHANGES_REQUESTED** [review]

File: pymc3/distributions/multivariate.py (hunk 1 above)
Comment (brandonwillard) [reply pymc3/distributions/multivariate.py:1579]: OK, then we don't need to add that functionality here/now.

File: pymc3/tests/test_distributions.py (hunk 8 above)
Comment (kc611) [reply pymc3/tests/test_distributions.py]: You were right, the passed all the way upto 4 decimal points in `float32` mode. And started showing failures at 5 decimal points. I left it at our default of 3 cause it may show flaky failures at 4 decimal points.

File: pymc3/distributions/multivariate.py (hunk 5 above)
Comment (kc611) [reply pymc3/distributions/multivariate.py]: The current approach should cover the correct `ndim_supp` value as well as the returned shape correctly. I simply returned the shape of `V` from the `param_shapes` list.

File: pymc3/tests/test_distributions_random.py [hunk 9]
```diff
@@ -1306,23 +1306,29 @@ class TestOrderedProbit(BaseTestDistribution):
...
+        (2, 4, 2, 3, 3),
+    ]
 
     reference_dist = lambda self: functools.partial(
         self.wishart_rng_fn, rng=self.get_random_state()
     )
     tests_to_run = [
         "check_rv_size",
         "check_pymc_params_match_rv_op",
-        "check_pymc_draws_match_reference",
```
Comment (ricardoV94) [inline pymc3/tests/test_distributions_random.py]: Why did you remove that test?

Comment (kc611) [reply pymc3/tests/test_distributions_random.py]: The test was failing with `AssertionError` with different numeric values, and this is flaky (see e4d10d0a636761339cea20a3d8eb822acbfd4b67 for a passing instance) and considering https://github.com/pymc-devs/pymc3/blob/7be2025c4091e450f9e9dd050eb0dca08995c83f/pymc3/tests/test_distributions_random.py#L1620 I though it would be better to remove it and just check for shapes. The test for `Wishart` in `test_distributions.py` is also commented out giving the reason https://github.com/pymc-devs/pymc3/blob/7be2025c4091e450f9e9dd050eb0dca08995c83f/pymc3/tests/test_distributions.py#L2043-L2047

(I don't know how applicable this is in the current scenario though)

Comment (ricardoV94) [reply pymc3/tests/test_distributions_random.py]: This is about the random method, not the `logp`/`logdf` as in the second test you linked to. If the scipy method is not good for testing, it is also not good enough to be used inside the `rng_fn`. Unless I am missing something.

Comment (kc611) [reply pymc3/tests/test_distributions_random.py]: Alright I re-added it. It showed a flaky failure (only once) on my system, if it arises again, we can investigate into this further.

Maybe it could've been caused by typecasting `nu` as `np.int(nu)` in `Wishart`'s `rng_fn`. I added a similar function in the test.

File: pymc3/tests/test_distributions_random.py [hunk 10]
```diff
@@ -1306,6 +1305,65 @@ class TestOrderedProbit(BaseTestDistribution):
...
+
+    pymc_dist_params = {"mu": mu, "rowcov": row_cov, "colcov": col_cov}
+    expected_rv_op_params = {"mu": mu, "rowcov": row_cov, "colcov": col_cov}
+
+    sizes_to_check = [None, (1), (2, 3)]
+    sizes_expected = [(3,), (1, 3), (2, 3, 3)]
+
+    tests_to_run = ["check_rv_size", "check_pymc_params_match_rv_op", "test_matrix_normal"]
+
+    def test_matrix_normal(self):
```
Comment (kc611) [inline pymc3/tests/test_distributions_random.py:1406]: Should we remove this test / replace this test with `rng_fn` from MatrixNormal ?

Comment (ricardoV94) [reply pymc3/tests/test_distributions_random.py:1406]: I think the current one is better, as it provides a more independent check that our `rng_fn` is doing the correct thing. The standard test is used when we are just relying on a standard implementation in `rng_fn` (i.e., just calling scipy / numpy, without many steps in it). Is the test passing?

Comment (kc611) [reply pymc3/tests/test_distributions_random.py:1406]: > Is the test passing?

It's flaky. Currently fails (in the current run of the Github jobs) but passed before.

File: pymc3/tests/test_distributions_random.py [hunk 11]
```diff
@@ -1306,6 +1305,65 @@ class TestOrderedProbit(BaseTestDistribution):
...
+    sizes_to_check = [None, (), 1, (1,), 5, (4, 5), (2, 4, 2)]
+    sizes_expected = [
+        (3, 3),
+        (3, 3),
+        (1, 3, 3),
+        (1, 3, 3),
+        (5, 3, 3),
+        (4, 5, 3, 3),
+        (2, 4, 2, 3, 3),
+    ]
```
Comment (brandonwillard) [inline pymc3/tests/test_distributions_random.py]: What extra logic (e.g. implemented in `WishartRV`) are all these sizes testing?  

Remember, we don't want the test time to grow larger without proportionally increasing the actual test coverage (and that coverage **doesn't** include NumPy/SciPy's code).  This test module is easily one of the most egregious violators of that principal, so we should always be a little more critical of additions to it.

Comment (kc611) [reply pymc3/tests/test_distributions_random.py]: Should that be checking only for edge cases like `None` and `()` ? I just took the `sizes_to_check` list from the original function in `BaseTestDistribution` where the sizes are being checked by default. (And constructed the corresponding expected list.)

Comment (ricardoV94) [reply pymc3/tests/test_distributions_random.py]: To echo @kc611, this is not actually testing more than what the default test does, just setting the specific parameters of the size test, which usually have to be adapted for multivariate distributions.

File: pymc3/tests/test_distributions_random.py (hunk 10 above)
Comment (ricardoV94) [reply pymc3/tests/test_distributions_random.py:1406]: You might need to check dimension by dimension. The default test flattens everything and that might not be so comparable. This issue also cropped in here: https://github.com/pymc-devs/pymc3/pull/4596#issuecomment-822589620

(I might also be completely off)

Comment (kc611) [reply pymc3/tests/test_distributions_random.py:1406]: Looks like that could work, I added it for now.

File: pymc3/tests/test_distributions_random.py [hunk 12]
```diff
@@ -1352,16 +1350,40 @@ class TestMatrixNormal(BaseTestDistribution):
...
+            check = pm.sample_prior_predictive(n_fails)
+
+        ref_smp = ref_rand(size[0], mu=np.random.random((3, 3)), rowcov=np.eye(3), colcov=np.eye(3))
+
+        p, f = delta, n_fails
+        while p <= delta and f > 0:
+            matrixnormal_smp = check["mvnormal"][f - 1, :, :]
+            curr_ref_smp = ref_smp[f - 1, :, :]
+
+            p = min(
```
Comment (ricardoV94) [inline pymc3/tests/test_distributions_random.py]: Not sure about the `min` here.

File: pymc3/tests/test_distributions_random.py [hunk 13]
```diff
@@ -1352,16 +1350,40 @@ class TestMatrixNormal(BaseTestDistribution):
     tests_to_run = ["check_rv_size", "check_pymc_params_match_rv_op", "test_matrix_normal"]
 
     def test_matrix_normal(self):
+        delta = 0.05  # limit for KS p-value
+        n_fails = 100  # Allows the KS fails a certain number of times
```
Comment (ricardoV94) [inline pymc3/tests/test_distributions_random.py]: 100 `n_fails` also seems extreme.

Comment (kc611) [reply pymc3/tests/test_distributions_random.py]: I set it to the default of 10.

File: pymc3/tests/test_distributions_random.py (hunk 12 above)
Comment (kc611) [reply pymc3/tests/test_distributions_random.py]: We can take an mean maybe ?

Comment (ricardoV94) [reply pymc3/tests/test_distributions_random.py]: Nevermind, `min` is the strongest test, I thought about it the wrong way. nfails can be a bit bigger than 10 but 100 seemed too much.

File: pymc3/distributions/multivariate.py [hunk 14]
```diff
@@ -737,6 +738,25 @@ def __str__(self):
...
+class WishartRV(RandomVariable):
+    name = "wishart"
+    ndim_supp = 2
+    ndims_params = [0, 2]
+    dtype = "floatX"
+    _print_name = ("Wishart", "\\operatorname{Wishart}")
+
+    def _shape_from_params(self, dist_params, rep_param_idx=1, param_shapes=None):
+        # The shape of second parameter `V` defines the shape of the output.
+        return param_shapes[1]
```
Comment (ricardoV94) [inline pymc3/distributions/multivariate.py]: Does this make sense like this @brandonwillard, I am not very familiar with the signature.

**Review by ricardoV94 — COMMENTED** [review]
LGTM otherwise. Is anything else missing?

File: pymc3/distributions/multivariate.py (hunk 14 above)
Comment (kc611) [reply pymc3/distributions/multivariate.py]: We can also use `dist_params[1].shape`. Just needs to be same as shape of second argument `V`.

File: pymc3/tests/test_distributions_random.py [hunk 15]
```diff
@@ -1306,6 +1303,91 @@ class TestOrderedProbit(BaseTestDistribution):
...
+            check = pm.sample_prior_predictive(n_fails)
+
+        ref_smp = ref_rand(size[0], mu=np.random.random((3, 3)), rowcov=np.eye(3), colcov=np.eye(3))
+
+        p, f = delta, n_fails
+        while p <= delta and f > 0:
+            matrixnormal_smp = check["mvnormal"][f - 1, :, :]
+            curr_ref_smp = ref_smp[f - 1, :, :]
+
+            p = np.mean(
```
Comment (ricardoV94) [inline pymc3/tests/test_distributions_random.py]: Sorry for the earlier confusion, but I think the minimum is the right metric to use. If the 10 attempts are not enough you can increase those a bit, but we should still use the min.

File: pymc3/tests/test_distributions.py [hunk 16]
```diff
@@ -2040,7 +2045,6 @@ def test_AR1(self, n):
         self.check_logp(AR1, Vector(R, n), {"k": Unit, "tau_e": Rplus}, AR1_logpdf)
 
     @pytest.mark.parametrize("n", [2, 3])
-    @pytest.mark.xfail(reason="Distribution not refactored yet")
     def test_wishart(self, n):
```
Comment (michaelosthege) [inline pymc3/tests/test_distributions.py:2048]: Over in #4824 I noticed that this test XPASS, but removing the XFAIL also doesn't make much sense, since the test has no active content.
Can you bring this test back online?

Comment (ricardoV94) [reply pymc3/tests/test_distributions.py:2048]: That tests requires the `checkd` which has not been refactored, and it seems the original test was no good either. I think the best is to just remove it.

Comment (kc611) [reply pymc3/tests/test_distributions.py:2048]: Updated it to the standard `check_logp` test.

File: pymc3/distributions/multivariate.py [hunk 17]
```diff
@@ -775,9 +796,10 @@ class Wishart(Continuous):
     This distribution is unusable in a PyMC3 model. You should instead
     use LKJCholeskyCov or LKJCorr.
     """
+    rv_op = wishart
 
-    def __init__(self, nu, V, *args, **kwargs):
-        super().__init__(*args, **kwargs)
+    @classmethod
+    def dist(cls, nu, V, *args, **kwargs):
         warnings.warn(
```
Comment (michaelosthege) [inline pymc3/distributions/multivariate.py:806]: Looks like @twiecki fixed the distribution in https://github.com/pymc-devs/pymc3/pull/701 ?
Does that mean we can remove the warning, or at least "downgrade" the warning into a recommendation to use LKJ dists instead?

**Disclaimer:** I have no clue what either distributions are 😅

File: pymc3/distributions/multivariate.py [hunk 18]
```diff
@@ -1533,175 +1565,101 @@ class MatrixNormal(Continuous):
...
-        self.solve_lower = solve_lower_triangular
-        self.solve_upper = solve_upper_triangular
-
-    def _setup_matrices(self, colcov, colchol, coltau, rowcov, rowchol, rowtau):
+
         cholesky = Cholesky(lower=True, on_error="raise")
 
+        if mu.ndim == 1:
+            raise ValueError(
+                "1x1 Matrix was provided, " " Please use Normal distribution " "for such cases."
```
Comment (michaelosthege) [inline pymc3/distributions/multivariate.py]:
```suggestion
                "1x1 Matrix was provided. Please use Normal distribution for such cases."
```

File: pymc3/distributions/multivariate.py [hunk 19]
```diff
@@ -1533,175 +1565,101 @@ class MatrixNormal(Continuous):
...
+            )
+
         # Among-row matrices
-        if len([i for i in [rowtau, rowcov, rowchol] if i is not None]) != 1:
+        if len([i for i in [rowcov, rowchol] if i is not None]) != 1:
             raise ValueError(
-                "Incompatible parameterization. "
-                "Specify exactly one of rowtau, rowcov, "
-                "or rowchol."
+                "Incompatible parameterization. " "Specify exactly one of rowcov, " "or rowchol."
```
Comment (michaelosthege) [inline pymc3/distributions/multivariate.py]:
```suggestion
                "Incompatible parameterization. Specify exactly one of rowcov, or rowchol."
```

File: pymc3/distributions/multivariate.py [hunk 20]
```diff
@@ -1533,175 +1565,101 @@ class MatrixNormal(Continuous):
...
+            rowchol_cov = at.as_tensor_variable(rowchol)
 
         # Among-column matrices
-        if len([i for i in [coltau, colcov, colchol] if i is not None]) != 1:
+        if len([i for i in [colcov, colchol] if i is not None]) != 1:
             raise ValueError(
-                "Incompatible parameterization. "
-                "Specify exactly one of coltau, colcov, "
-                "or colchol."
+                "Incompatible parameterization. " "Specify exactly one of colcov, " "or colchol."
```
Comment (michaelosthege) [inline pymc3/distributions/multivariate.py]:
```suggestion
                "Incompatible parameterization. Specify exactly one of colcov, or colchol."
```

File: pymc3/tests/test_distributions_random.py (hunk 11 above)
Comment (michaelosthege) [reply pymc3/tests/test_distributions_random.py]: I agree to both: This implementations is very much in line with the defaults.
The the `BaseTestDistribution` for sure takes a sledgehammer to break a nut, but that problem is out of scope of this PR.

**Review by michaelosthege — COMMENTED** [review]
@kc611 take a look at my comments and tell us when you see this as ready to merge :)

File: pymc3/tests/test_distributions_random.py (hunk 11 above)
Comment (brandonwillard) [reply pymc3/tests/test_distributions_random.py]: Pushing admittedly poor defaults forward is a great way to get into a bad situation, and an even better way to get deeper into existing ones.  

Also, the answer is not one big PR that removes all the testing redundancies; given the knowledge and effort required for that, it's quite unrealistic.  Plus, that won't prevent the problem from reappearing again and again after such a heroic PR.

At the very least, we need to stop adding to the testing redundancies on a *case-by-case basis*, and we can do that by looking for it during reviews; otherwise, we're just hoping that contributors will do that for us, and that's not how things have turned out.


Regarding these exact tests, if they're only targeting `WishartRV.rng_fn`, it looks like we only need a test for an empty and non-empty `size` (and perhaps one other), because that's the only relevant condition in the code added by this PR.  Every other value besides the first two appears to be testing the code in `RandomVariable` and `scipy.stats.wishart.rvs`.

Comment (kc611) [reply pymc3/tests/test_distributions_random.py]: I think there's no harm in not testing the `[ (), (1,), 5,  (2, 4, 2)]` cases if we are testing for `[None, 1, (4, 5)]` cases since they're pretty much the same thing with different values. (Updated accordingly)

File: pymc3/distributions/multivariate.py (hunk 17 above)
Comment (kc611) [reply pymc3/distributions/multivariate.py:806]: I think #701 adds the `WishartBartlett` as a solution for that, I updated the warning to recommend use of  `WishartBartlett` instead of `LKJCorr` or `LKJCholeskyCov`.

cc @twiecki Is this correct ?

Comment (twiecki) [reply pymc3/distributions/multivariate.py:806]: No, LKJCorr should be used. `Wishart` itself is basically unusable, so if a user really wants that distribution, they can only use `WishartBartlett`, but they should always use either `LKJCorr` or `LKJCholeskyCov`.

Comment (ricardoV94) [reply pymc3/distributions/multivariate.py:806]: Let's revert the warning to how it was before? Other than this PR looks pretty ready.

Also, should this warn be in `__new__`? Does it appear twice during a normal initialization inside a `pm.Model()`?

**Review by ricardoV94 — APPROVED** [review]
LGTM, let's wait for the tests. Great work, these two were a tough group to crack

File: pymc3/distributions/multivariate.py (hunk 17 above)
Comment (michaelosthege) [reply pymc3/distributions/multivariate.py:806]: I don't think it would appear twice. And implementing a `__new__` just because of this warning seems a little too much.
Let's just rephrase to mention the entire recommendation from above and then go for it.

Comment (michaelosthege) [reply pymc3/distributions/multivariate.py:806]: Or like this. (I changed as I was writing the above.)

**Review by michaelosthege — APPROVED** [review]
